uccb/src/bin/validator-oracle.rs

```diff
@@ -0,0 +1,273 @@
+                contracts::deposit::ABI.clone(),
+            )?)?,
+            validator_manager_abi: serde_json::from_value(
+                serde_json::from_str::<serde_json::Value>(VALIDATOR_MANAGER_ABI_JSON)?["abi"]
+                    .clone(),
+            )?,
+        })
+    }
+
+    pub async fn start(&mut self) -> Result<()> {
+        info!(
+            "Starting validator oracle with signer address {}",
+            self.signer.address()
+        );
+
+        let validators = self.get_stakers().await?;
+        info!("Current validator set is: {}", Display(&validators));
+
+        let (sender, receiver) = watch::channel(validators);
+
+        let mut handles = vec![];
+        for chain_client in &self.chain_clients {
+            let mut receiver = receiver.clone();
+            let chain_client = chain_client.clone();
+            let validator_manager_abi = self.validator_manager_abi.clone();
+            let handle = tokio::spawn(async move {
+                loop {
+                    let validators = receiver.borrow_and_update().clone();
+                    if Self::update_validator_manager(
+                        &chain_client,
+                        &validators,
+                        validator_manager_abi.clone(),
+                    )
+                    .await
+                    .is_err()
+                    {
+                        error!(
+                            "Failed updating the validator manager on {}",
+                            chain_client.rpc_url
+                        );
+                    }
+
+                    if receiver.changed().await.is_err() {
+                        break;
+                    }
+                }
+            });
+            handles.push(handle);
+        }
+
+        let result = self.listen_to_staker_updates(sender).await;
+
+        for handle in handles {
+            handle.await?;
+        }
+
+        result
+    }
+
+    async fn listen_to_staker_updates(
+        &mut self,
+        sender: watch::Sender<Vec<Address>>,
+    ) -> Result<()> {
+        let filter = Filter::new()
+            .address(contract_addr::DEPOSIT)
+            // Must be the same signature as the event in deposit.sol
+            .event("StakerAdded(bytes)")
+            .from_block(BlockNumberOrTag::Finalized);
+
+        let subscription = self
+            .zq2_chain_client()
+            .provider
+            .subscribe_logs(&filter)
+            .await?;
+        let mut stream = subscription.into_stream();
+        while let Some(log) = stream.next().await {
+            // TODO: infer if staker added or removed
+            info!("Received validator update: {log:?}");
+
+            let validators = self.get_stakers().await?;
+            info!(
+                "Updating chains to the current validator set to: {}",
+                Display(&validators)
+            );
+
+            sender.send(validators)?;
```

Comment:
I believe we do. Plus, it's easier setting it on the filter than writing the logic of what you suggested.

Reply:
Where? I looked and couldn't find anywhere we do. You specify `from_block(BlockNumberOrTag::Finalized)` in the subscription filter, but all this means is that the subscription should return blocks greater than or equal to the finalized height *at the time* the subscription was created.

Reply:
... so? isn't that what we're interested in?

Reply:
Nope. We need to be sure each time the validator set is updated that the change is finalised. We can't update the oracles on other networks until we're 100% sure the change won't get rolled back on the ZQ2 network.
cc @DrZoltanFazekas

Reply:
That's why we specified the 'finalized' on the filter? A bit confused then.

Reply:
`eth_subscribe` lets us get a continuous stream of logs matching the specified filter. The `from_block = "finalized"` that we pass says "start the stream of logs at the **currently finalized** block". This will have the effect that it immediately yields logs from the last ~2 blocks. After this point, it will continue to stream logs from the latest blocks as they are mined, but before they are finalized. We need a separate check for finalization.

Reply:
What we want are only events with `to_block = "finalized"` i.e. events in blocks before or in a finalized block and not `from_block = "finalized"`. I wasn't sure if using the "finalized" tag in a subscription means that the respective block number is updated as newer blocks get finalized over time (I think this is what @yaron-zilliqa expected) or the block number is set to the current finalized block when the subscription is created and not updated afterwards (the behavior @JamesHinshelwood described). So I tested it with a surprising result:

```
curl -X POST -H "Content-Type: application/json" --data '{"jsonrpc":"2.0", "id": 1, "method":"eth_getBlockByNumber","params": ["finalized", false]}' https://api.zq2-prototestnet.zilliqa.com --silent | jq '.result.number' && wscat -c wss://api.zq2-prototestnet.zilliqa.com -x '{"jsonrpc":"2.0", "id": 1, "method":"eth_subscribe","params": ["logs", {"address": "0x9F56A6Ef4d8dc85164BCaF38b760553181B6fb8c", "toBlock": "finalized"}]}' && curl -X POST -H "Content-Type: application/json" --data '{"jsonrpc":"2.0", "id": 1, "method":"eth_getBlockByNumber","params": ["finalized", false]}' https://api.zq2-prototestnet.zilliqa.com --silent | jq '.result.number'
```
queried the finalized block number `0x1a1197` before and `0x1a119a` after the event was fired as well as the event in between:
```
"0x1a1197"
{"jsonrpc":"2.0","result":"0xa9b28a4ebdb17227","id":1}
{"jsonrpc":"2.0","method":"eth_subscription","params":{"subscription":"0xa9b28a4ebdb17227","result":{"address":"0x9f56a6ef4d8dc85164bcaf38b760553181b6fb8c","topics":["0x63a242a632efe33c0e210e04e4173612a17efa4f16aa4890bc7e46caece80de0"],"data":"0x0000000000000000000000000000000000000000000000000000000000000013","blockHash":"0x892d2a0deb3d732794ce81e06a2618f611f436d98fd739f2f5f46cad4c4ecf04","blockNumber":"0x1a119b","blockTimestamp":"0x6700264c","transactionHash":"0x9dd17f74aedb8e10166580ec05404e820858036f8e58244d125f8e7ef4dd8254","transactionIndex":"0x0","logIndex":"0x0","removed":false}}}
"0x1a119a"
```
The event in block `0x1a119b` was returned by the subscription although that block was definitely not finalized yet. The finalized block number after that event had been returned was only `0x1a119a`.

I thought it may be a bug in ZQ2 so I tried to reproduce the behavior on Holesky and BSC, but both networks returned an error when I was using the "finalized" tag in log subscriptions:
```
{"jsonrpc":"2.0","error":{"code":-32000,"message":"invalid block range params"},"id":1}
```
Interestingly, the "latest" tag worked, but "finalized" was not rejected.